File: Data Structures & Algorithms/generate-parentheses/submission-3.cpp
class Solution {
public:
vector<string> generateParenthesis(int n)
    {
        string s = "";
        vector<string> results;
        dfs(s, 0, 0, results, n);
        return results;
    }

    void dfs(string s, int left, int right, vector<string> &results, int n)
    {
        if (s.size() == 2 * n)
        {
            results.push_back(s);
        }

        if (left < n)
        {
            dfs(s + "(", left + 1, right, results, n);
        }
        if (right < left)
        {
            dfs(s + ")", left, right + 1, results, n);
        }
    }
};

File: Data Structures & Algorithms/bitwise-and-of-numbers-range/submission-3.py
class Solution:
    def rangeBitwiseAnd(self, left: int, right: int) -> int:
        if left == right:
            return left
        msbIndexLeft = left.bit_length() - 1
        msbIndexRight = right.bit_length() - 1

        total = 0
        while msbIndexLeft > 0:
            mask = 1 << msbIndexRight
            msbIndexRight -= 1
            msbIndexLeft -= 1
            lmask = left & mask
            rmask = right & mask
            if lmask != rmask:
                break
            total |= lmask
        return total